cpp/include/cudf/reduction.hpp

```diff
@@ -0,0 +1,107 @@
+};
+
+/**
+ * @brief These enums indicate the supported reduction operations that can be
+ * performed on a column
+ */
+enum class reduction_op {
+  SUM = 0,        ///< Computes the sum of all values in the column
+  MIN,            ///< Computes the minimum of all values in the column
+  MAX,            ///< Computes the maximum of all values in the column
+  PRODUCT,        ///< Computes the multiplicative product of all values in the column
+  SUMOFSQUARES,   ///< Computes the sum of squares of the values in the column
+  MEAN,           ///< Computes the arithmetic mean of the values in the column
+  VAR,            ///< Computes the variance of the values in the column
+  STD,            ///< Computes the standard deviation of the values in the column
+  ANY,            ///< Computes to true if any of the values are non-zero/true
+  ALL,            ///< Computes to true if all of the values are non-zero/true
+};
+
+/** --------------------------------------------------------------------------*
+ * @brief  Computes the reduction of the values in all rows of a column.
+ * This function does not detect overflows in reductions.
+ * Using a higher precision `data_type` may prevent overflow.
+ * Only `min` and `max` ops are supported for reduction of non-arithmetic
+ * types (timestamp, string...).
+ * The null values are skipped for the operation.
+ * If the column is empty, the member `is_valid()` of the output scalar
+ * will contain `false`.
+ *
+ * @throws `cudf::logic_error` if reduction is called for non-arithmetic output
+ * type and operator other than `min` and `max`.
+ * @throws `cudf::logic_error` if input column data type is not convertible to
+ * output data type.
+ * If the input column has arithmetic type, output_dtype can be any arithmetic
+ * type. For `mean`, `var` and `std` ops, a floating point output type must be 
+ * specified. If the input column has non-arithmetic type
+ *   eg.(timestamp, string...), the same type must be specified.
+ *
+ * @param[in] col Input column view
+ * @param[in] op  The operator applied by the reduction
+ * @param[in] output_dtype  The computation and output precision.
+ * @params[in] mr The resource to use for all allocations
+ * @param[in] ddof Delta Degrees of Freedom: the divisor used in calculation of
+ * `std` and `var` is `N - ddof`, where `N` is the population size.`
+ * @returns  cudf::scalar the result value
+ * If the reduction fails, the member is_valid of the output scalar
+ * will contain `false`.
+ * ----------------------------------------------------------------------------**/
+std::unique_ptr<scalar> reduce(
+    const column_view& col, reduction_op op, data_type output_dtype,
+    cudf::size_type ddof = 1,
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/** --------------------------------------------------------------------------*
+ * @brief  Computes the scan (a.k.a. prefix sum) of a column.
```

Comment:
It's not just a sum. 
```suggestion
 * @brief  Computes the scan of a column.
```

Reply:
FYI, it's historically common to refer to scan as prefix sum, regardless of the binary operator used.  I think this "aka" was actually helpful since `scan` can be ambiguous.